### PR TITLE
shape: Free platform-specific shaped window data.

### LINKED ISSUE #2128: Shaped windows creation and destruction
# This bug report was migrated from our old Bugzilla tracker.

**Reported in version:** HG 2.1
**Reported for operating system, platform:** All, All

# Comments on the original bug report:

On 2016-03-15 18:22:10 +0000, Juha Niemimäki wrote:

> There might be a couple of issues with shaped windows:
> 
> 1) SDL doesn't check the function pointers (like CreateShaper) before using ( https://hg.libsdl.org/SDL/file/9cec5fe32bca/src/video/SDL_shape.c#l38 )
> 
> 2) Cleanup: shouldn't shaper and data be cleaned during window destruction? Perhaps SDL_DestroyWindow could check for shaper existance, and when found, call a shaper destructor function?

### PR DESCRIPTION
**Pull request by icculus**
Fixes #2128.

Just letting this sit in a pull request so I can make sure I didn't break the build before pushing.

### CODE REVIEW

File: src/video/windows/SDL_windowsshape.c [hunk 1]
```diff
@@ -37,10 +37,12 @@ Win32_CreateShaper(SDL_Window * window) {
     result->driverdata = (SDL_ShapeData*)SDL_malloc(sizeof(SDL_ShapeData));
     ((SDL_ShapeData*)result->driverdata)->mask_tree = NULL;
     /* Put some driver-data here. */
-    window->shaper = result;
```
Comment (sezero) [inline src/video/windows/SDL_windowsshape.c:40]: Did you mean to move this before `return result;` ?

Comment (sezero) [reply src/video/windows/SDL_windowsshape.c:40]: I think the following is needed, on top of existing patch: `Win32_ResizeWindowShape`
assumes a legitimate shaper pointer.
```diff
diff --git a/src/video/windows/SDL_windowsshape.c b/src/video/windows/SDL_windowsshape.c
index 24f125b..7d15f4d 100644
--- a/src/video/windows/SDL_windowsshape.c
+++ b/src/video/windows/SDL_windowsshape.c
@@ -39,2 +39,3 @@ Win32_CreateShaper(SDL_Window * window) {
     /* Put some driver-data here. */
+    window->shaper = result;
     resized_properly = Win32_ResizeWindowShape(window);
@@ -43,2 +44,3 @@ Win32_CreateShaper(SDL_Window * window) {
         SDL_free(result);
+        window->shaper = NULL;
         return NULL;
```

Comment (icculus) [reply src/video/windows/SDL_windowsshape.c:40]: The caller of CreateShaper assigns the return value to window->shaper.

Comment (icculus) [reply src/video/windows/SDL_windowsshape.c:40]: Oh wait, I see, you're right.

Comment (icculus) [reply src/video/windows/SDL_windowsshape.c:40]: ok, forced-pushed this correction, thanks!

**Review by sezero — APPROVED** [review]

**Pull request merged**